knowledge-base/adr/0021-odata-url-builder.md

```diff
@@ -0,0 +1,193 @@
+# Title
+Status: in progress
+
+Decision: ??
+
+## Background
+### How to build a OData request by using the SDK
+So far, the SDK user can build a request with the pattern like:
+``` ts
+People.reqeustBuilder()
+  .someOperation(params) // getAll(), getByKey(...keys), create(entity), update(entity), delete(entity | ...keys)
+```
+which then generates url like below that has only **ONE** level:
+```
+/servicePath/People // getAll(), create, update, delete(entity)
+/servicePath/People(1) // getByKey, delete(key)
+```
+
+### Exception: `asChildOf`
+The API `asChildOf` is designed for generating a URL that contains **TWO** levels (entity + navigation property).
+The code snippet below:
+``` ts
+Friends.requestBuilder()
+  .create(newFriend)
+  .asChildOf( 
+    person, // for getting the key
+    People.Friends
+)
+```
+will generate url like:
+```
+/People(personKey)/Friends
+```
+
+This is a bit confusing, because the order of the ts code is from navigation property to parent entity, while being reversed in the generated URL.
+As a workaround, it might be still fine to keep the same API structure.
+
+### Problem
+There are still lots of OData requests that are not supported by the SDK:
+1. [high prio] `GET /People('scottketchum')/Friends` for querying navigation properties (1-to-1 and 1-to-many), requested by users
+2. [high prio] `GET /People('scottketchum')/Friends('russellwhyte')`, for querying 1-to-many navigation properties with keys.
+3. [high prio] `GET /People('russellwhyte')/BestFriend/BestFriend/BestFriend`, infinite links
+4. [high prio] `GET /People('scottketchum')/Friends('russellwhyte')/BestFriend`, similar case with 1-to-many navigation properties
+5. [medium prio] `GET /People('scottketchum')/$value`, requested by users and used for fetching the spec files with ["APIContent service"](https://github.tools.sap/cloudsdk/cloud-sdk-api-mirror/blob/main/fetch-spec-from-api-hub/http-calls.ts#L136) ([odata doc](https://www.odata.org/getting-started/basic-tutorial/))
+6. [low prio] `GET /People('scottketchum')/Friends/$count`, count the number of friend.
+7. [low prio] `GET /People('scottketchum')/FirstName`, similar to the navigation properties
+
+### Focus
+We will focus on the problems 1-4 for now and might consider rough ideas for 5-7 with extension.
+
+## Alternatives
+### Proposal A (Consider as separate task as it might be used as workaround for all cases)
+A new API
+```
+People.reqeustBuilder()
+  .getByKey(key)
+  .appendUrl('$value')
+  // .appendUrl('Friends')
+  // .appendUrl("Friends('russellwhyte')")
```

Comment:
`.appendPath()`

Reply:
Please don't forget to adjust ;)

Reply:
Updated, as the api name seems to be aligned.